server/src/main/java/org/elasticsearch/cluster/coordination/JoinRequest.java

```diff
@@ -56,6 +69,9 @@ public JoinRequest(StreamInput in) throws IOException {
     public void writeTo(StreamOutput out) throws IOException {
         super.writeTo(out);
         sourceNode.writeTo(out);
+        if (out.getTransportVersion().onOrAfter(TransportVersion.V_8_8_0)) {
+            TransportVersion.writeVersion(out, transportVersion);
```

Comment:
Why do we need to write the transport version to the stream when the stream already has the transport version attached? ie when could this version be different than that from the stream?

Reply:
If the joining node is running a newer version than the master it's joining then the version on the stream will be the master's `TransportVersion#CURRENT`, not the one for the joining node.

Reply:
Ahh, I understand now. Thanks!

Reply:
(maybe this is a good illustration of my concern about bugs in this area - it's subtle and tricksy even when we're thinking hard about it :grin:)